File: scripts/gitea_api.py
"""
Shared Gitea API helper functions used by multiple scripts.
"""

import os
import requests
from config import GITEA_URL


def get_session(admin_user, admin_pass):
    """Create an authenticated requests session."""
    session = requests.Session()
    session.auth = (admin_user, admin_pass)
    return session


def user_exists(session, username):
    """Check if a user exists."""
    resp = session.get(f"{GITEA_URL}/api/v1/users/{username}")
    return resp.status_code == 200


def get_user_orgs(session, username):
    """Get list of organization names a user belongs to."""
    orgs = []
    page = 1
    while True:
        resp = session.get(
            f"{GITEA_URL}/api/v1/users/{username}/orgs",
            params={"page": page, "limit": 50},
        )
        if resp.status_code != 200:
            return None
        data = resp.json()
        if not data:
            break
        orgs.extend(o["username"] for o in data)
        page += 1
    return orgs


# --- Teams ---

def find_team(session, course_name, team_name):
    """Find an existing team by name in a course."""
    page = 1
    while True:
        resp = session.get(
            f"{GITEA_URL}/api/v1/orgs/{course_name}/teams",
            params={"page": page, "limit": 50},
        )
        if resp.status_code != 200:
            return None
        teams = resp.json()
        if not teams:
            return None
        for team in teams:
            if team.get("name") == team_name:
                return team
        page += 1


def get_all_teams(session, course_name, include_owners=False):
    """Get all teams in a course. Returns list of team dicts."""
    teams = []
    page = 1
    while True:
        resp = session.get(
            f"{GITEA_URL}/api/v1/orgs/{course_name}/teams",
            params={"page": page, "limit": 50},
        )
        if resp.status_code != 200:
            break
        data = resp.json()
        if not data:
            break
        for t in data:
            if include_owners or t["name"] != "Owners":
                teams.append(t)
        page += 1
    return teams


def get_org_teams_dict(session, course_name):
    """Get all teams in a course as dict {team_name: team_data}."""
    teams = {}
    page = 1
    while True:
        resp = session.get(
            f"{GITEA_URL}/api/v1/orgs/{course_name}/teams",
            params={"page": page, "limit": 50},
        )
        if resp.status_code != 200:
            return teams
        data = resp.json()
        if not data:
            break
        for t in data:
            teams[t["name"]] = t
        page += 1
    return teams


def ensure_team(session, course_name, team_name, permission="write"):
    """Create a team or use existing one, verifying settings."""
    expected_units = ["repo.code", "repo.issues", "repo.pulls"]

    resp = session.post(
        f"{GITEA_URL}/api/v1/orgs/{course_name}/teams",
        json={
            "name": team_name,
            "permission": permission,
            "includes_all_repositories": False,
            "units": expected_units,
        },
    )
    if resp.status_code == 201:
        return "created", resp.json()

    team = find_team(session, course_name, team_name)
    if not team:
        return "error", resp.text

    issues = []
    if team.get("permission") != permission:
        issues.append(f"permission: {team.get('permission')} -> {permission}")
    if team.get("includes_all_repositories"):
        issues.append("includes_all_repositories: true -> false")
    current_units = sorted(team.get("units", []))
    if current_units != sorted(expected_units):
        issues.append(f"units: {current_units} -> {sorted(expected_units)}")

    if issues:
        patch = session.patch(
            f"{GITEA_URL}/api/v1/teams/{team['id']}",
            json={
                "permission": permission,
                "includes_all_repositories": False,
                "units": expected_units,
            },
        )
        if patch.status_code == 200:
            print(f"  FIXED: Team settings updated ({', '.join(issues)}).")
            team = patch.json()
        else:
            print(f"  WARN: Could not update team settings (HTTP {patch.status_code})")

    return "exists", team


def get_team_members(session, team_id):
    """Get all member logins of a team."""
    members = []
    page = 1
    while True:
        resp = session.get(
            f"{GITEA_URL}/api/v1/teams/{team_id}/members",
            params={"page": page, "limit": 50},
        )
        if resp.status_code != 200:
            break
        data = resp.json()
        if not data:
            break
        members.extend(m["login"] for m in data)
        page += 1
    return members


def get_team_repos(session, team_id):
    """Get repo names assigned to a team."""
    repos = []
    page = 1
    while True:
        resp = session.get(
            f"{GITEA_URL}/api/v1/teams/{team_id}/repos",
            params={"page": page, "limit": 50},
        )
        if resp.status_code != 200:
            break
        data = resp.json()
        if not data:
            break
        repos.extend(r["name"] for r in data)
        page += 1
    return repos


def is_team_member(session, team_id, username):
    """Check if a user is a member of a team."""
    resp = session.get(
        f"{GITEA_URL}/api/v1/teams/{team_id}/members/{username}",
    )
    return resp.status_code == 200


def add_team_member(session, team_id, username):
    """Add a user to a team."""
    resp = session.put(
        f"{GITEA_URL}/api/v1/teams/{team_id}/members/{username}",
    )
    return resp.status_code, resp.text


def remove_team_member(session, team_id, username):
    """Remove a user from a team."""
    resp = session.delete(
        f"{GITEA_URL}/api/v1/teams/{team_id}/members/{username}",
    )
    return resp.status_code, resp.text


def add_team_repo(session, team_id, course_name, repo_name):
    """Assign a repo to a team."""
    resp = session.put(
        f"{GITEA_URL}/api/v1/teams/{team_id}/repos/{course_name}/{repo_name}",
    )
    return resp.status_code, resp.text


# --- Repos ---

def create_blank_repo(session, owner, repo_name):
    """Create a blank private repo under a course."""
    resp = session.post(
        f"{GITEA_URL}/api/v1/orgs/{owner}/repos",
        json={
            "name": repo_name,
            "private": True,
            "auto_init": True,
        },
    )
    if resp.status_code == 201:
        return "created", resp.json()
    return "error", resp.text


def ensure_repo(session, template_owner, template_repo, new_owner, new_repo):
    """Create a repo from template, or verify existing one is private."""
    resp = session.get(f"{GITEA_URL}/api/v1/repos/{new_owner}/{new_repo}")
    if resp.status_code == 200:
        repo = resp.json()
        if not repo.get("private"):
            patch = session.patch(
                f"{GITEA_URL}/api/v1/repos/{new_owner}/{new_repo}",
                json={"private": True},
            )
            if patch.status_code == 200:
                print(f"  FIXED: Repo '{new_owner}/{new_repo}' set to private.")
            else:
                print(f"  WARN: Could not set repo to private (HTTP {patch.status_code})")
        return "exists", repo
    resp = session.post(
        f"{GITEA_URL}/api/v1/repos/{template_owner}/{template_repo}/generate",
        json={
            "owner": new_owner,
            "name": new_repo,
            "private": True,
            "git_content": True,
            "topics": True,
            "labels": True,
        },
    )
    if resp.status_code == 201:
        return "created", resp.json()
    return "error", resp.text


def get_org_repos(session, course_name):
    """Get all repos in a course."""
    repos = []
    page = 1
    while True:
        resp = session.get(
            f"{GITEA_URL}/api/v1/orgs/{course_name}/repos",
            params={"page": page, "limit": 50},
        )
        if resp.status_code != 200:
            break
        data = resp.json()
        if not data:
            break
        repos.extend(data)
        page += 1
    return repos


# --- CSV file resolution ---

PROJECT_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))


def resolve_csv(course_name):
    """Resolve course name to CSV file path: {course_name}.csv in the project root."""
    return os.path.join(PROJECT_DIR, f"{course_name}.csv")


# --- CSV parsing ---

def parse_csv(file_path):
    """Parse CSV. Returns list of (course_name, student_id, team_name_or_None).

    CSV format: course_name student_id name [team_name]
    """
    entries = []
    with open(file_path, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            parts = line.split()
            course_name = parts[0]
            student_id = parts[1]
            team_name = parts[3] if len(parts) >= 4 else None
            entries.append((course_name, student_id, team_name))
    return entries


def parse_csv_teams(file_path):
    """Parse CSV into course name and teams for repo creation.

    CSV format: course_name student_id name team_name
    Returns (course_name, OrderedDict {team_name: [student_ids]}).
    """
    import sys
    from collections import OrderedDict

    course_names = set()
    teams = OrderedDict()

    with open(file_path, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            parts = line.split()
            if len(parts) < 4:
                continue
            course_names.add(parts[0])
            teams.setdefault(parts[3], []).append(parts[1])

    if len(course_names) > 1:
        print(f"Error: Multiple courses found in CSV: {', '.join(sorted(course_names))}")
        print("  All rows must belong to the same course.")
        sys.exit(1)

    if not course_names:
        return None, teams

    return next(iter(course_names)), teams


def validate_single_course(entries):
    """Validate all entries belong to the same course. Returns course name or exits."""
    import sys
    course_names = set(course for course, _, _ in entries)
    if len(course_names) > 1:
        print(f"Error: Multiple courses found in CSV: {', '.join(sorted(course_names))}")
        print("  All rows must belong to the same course.")
        sys.exit(1)
    return next(iter(course_names))

File: scripts/gitea_api.ts
/**
 * Shared Gitea API helper functions used by multiple scripts.
 */

import path from 'path';
import fs from 'fs';
import { fileURLToPath } from 'url';
import { GITEA_URL } from './config.ts';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
export const PROJECT_DIR = path.dirname(__dirname);
export const COURSES_DIR = path.join(PROJECT_DIR, 'courses');

export type CsvEntry = [string, string, string | null]; // [course_name, student_id, team_name]

export interface Session {
  admin_user: string;
  admin_pass: string;
  get(url: string, params?: Record<string, unknown>): Promise<Response>;
  post(url: string, body?: unknown): Promise<Response>;
  put(url: string, body?: unknown): Promise<Response>;
  patch(url: string, body?: unknown): Promise<Response>;
  delete(url: string): Promise<Response>;
}

interface ApiOpts {
  admin_user: string;
  admin_pass: string;
  body?: unknown;
  params?: Record<string, unknown>;
}

function auth_headers(admin_user: string, admin_pass: string): Record<string, string> {
  return { Authorization: 'Basic ' + Buffer.from(`${admin_user}:${admin_pass}`).toString('base64') };
}

async function api(method: string, url: string, opts: ApiOpts): Promise<Response> {
  const u = new URL(url, GITEA_URL);
  if (opts.params) for (const [k, v] of Object.entries(opts.params)) u.searchParams.set(k, String(v));
  const fetchOpts: RequestInit = {
    method,
    headers: { ...auth_headers(opts.admin_user, opts.admin_pass), 'Content-Type': 'application/json' },
  };
  if (opts.body) fetchOpts.body = JSON.stringify(opts.body);
  return fetch(u.href, fetchOpts);
}

export function get_session(admin_user: string, admin_pass: string): Session {
  const creds = { admin_user, admin_pass };
  return {
    ...creds,
    get: (url, params) => api('GET', url, { ...creds, params }),
    post: (url, body) => api('POST', url, { ...creds, body }),
    put: (url, body) => api('PUT', url, { ...creds, body }),
    patch: (url, body) => api('PATCH', url, { ...creds, body }),
    delete: (url) => api('DELETE', url, creds),
  };
}

export async function user_exists(session: Session, username: string): Promise<boolean> {
  const resp = await session.get(`${GITEA_URL}/api/v1/users/${username}`);
  return resp.status === 200;
}

export async function get_user_orgs(session: Session, username: string): Promise<string[] | null> {
  const orgs: string[] = [];
  let page = 1;
  while (true) {
    const resp = await session.get(`${GITEA_URL}/api/v1/users/${username}/orgs`, { page, limit: 50 });
    if (resp.status !== 200) return null;
    const data = await resp.json();
    if (!data.length) break;
    orgs.push(...data.map((o: { username: string }) => o.username));
    page++;
  }
  return orgs;
}

// --- Teams ---

export async function find_team(session: Session, course_name: string, team_name: string): Promise<Record<string, unknown> | null> {
  let page = 1;
  while (true) {
    const resp = await session.get(`${GITEA_URL}/api/v1/orgs/${course_name}/teams`, { page, limit: 50 });
    if (resp.status !== 200) return null;
    const teams = await resp.json();
    if (!teams.length) return null;
    const found = teams.find((t: { name: string }) => t.name === team_name);
    if (found) return found;
    page++;
  }
}

export async function get_all_teams(session: Session, course_name: string, include_owners = false): Promise<Record<string, unknown>[]> {
  const teams: Record<string, unknown>[] = [];
  let page = 1;
  while (true) {
    const resp = await session.get(`${GITEA_URL}/api/v1/orgs/${course_name}/teams`, { page, limit: 50 });
    if (resp.status !== 200) break;
    const data = await resp.json();
    if (!data.length) break;
    for (const t of data) {
      if (include_owners || t.name !== 'Owners') teams.push(t);
    }
    page++;
  }
  return teams;
}

export async function get_org_teams_dict(session: Session, course_name: string): Promise<Record<string, Record<string, unknown>>> {
  const teams: Record<string, Record<string, unknown>> = {};
  let page = 1;
  while (true) {
    const resp = await session.get(`${GITEA_URL}/api/v1/orgs/${course_name}/teams`, { page, limit: 50 });
    if (resp.status !== 200) return teams;
    const data = await resp.json();
    if (!data.length) break;
    for (const t of data) teams[t.name as string] = t;
    page++;
  }
  return teams;
}

export async function ensure_team(session: Session, course_name: string, team_name: string, permission = 'write'): Promise<[string, unknown]> {
  const expected_units = ['repo.code', 'repo.issues', 'repo.pulls'];

  const resp = await session.post(`${GITEA_URL}/api/v1/orgs/${course_name}/teams`, {
    name: team_name, permission, includes_all_repositories: false, units: expected_units,
  });
  if (resp.status === 201) return ['created', await resp.json()];

  const team = await find_team(session, course_name, team_name) as Record<string, unknown> | null;
  if (!team) return ['error', await resp.text()];

  const issues: string[] = [];
  if (team.permission !== permission) issues.push(`permission: ${team.permission} -> ${permission}`);
  if (team.includes_all_repositories) issues.push('includes_all_repositories: true -> false');
  const current_units = ((team.units as string[]) || []).slice().sort();
  if (JSON.stringify(current_units) !== JSON.stringify(expected_units.slice().sort())) {
    issues.push(`units: ${JSON.stringify(current_units)} -> ${JSON.stringify(expected_units.slice().sort())}`);
  }

  if (issues.length) {
    const patch = await session.patch(`${GITEA_URL}/api/v1/teams/${team.id}`, {
      permission, includes_all_repositories: false, units: expected_units,
    });
    if (patch.status === 200) {
      console.log(`  FIXED: Team settings updated (${issues.join(', ')}).`);
      return ['exists', await patch.json()];
    } else {
      console.log(`  WARN: Could not update team settings (HTTP ${patch.status})`);
    }
  }

  return ['exists', team];
}

export async function get_team_members(session: Session, team_id: number): Promise<string[]> {
  const members: string[] = [];
  let page = 1;
  while (true) {
    const resp = await session.get(`${GITEA_URL}/api/v1/teams/${team_id}/members`, { page, limit: 50 });
    if (resp.status !== 200) break;
    const data = await resp.json();
    if (!data.length) break;
    members.push(...data.map((m: { login: string }) => m.login));
    page++;
  }
  return members;
}

export async function get_team_repos(session: Session, team_id: number): Promise<string[]> {
  const repos: string[] = [];
  let page = 1;
  while (true) {
    const resp = await session.get(`${GITEA_URL}/api/v1/teams/${team_id}/repos`, { page, limit: 50 });
    if (resp.status !== 200) break;
    const data = await resp.json();
    if (!data.length) break;
    repos.push(...data.map((r: { name: string }) => r.name));
    page++;
  }
  return repos;
}

export async function is_team_member(session: Session, team_id: number, username: string): Promise<boolean> {
  const resp = await session.get(`${GITEA_URL}/api/v1/teams/${team_id}/members/${username}`);
  return resp.status === 200;
}

export async function add_team_member(session: Session, team_id: number, username: string): Promise<[number, string]> {
  const resp = await session.put(`${GITEA_URL}/api/v1/teams/${team_id}/members/${username}`);
  return [resp.status, await resp.text()];
}

export async function remove_team_member(session: Session, team_id: number, username: string): Promise<[number, string]> {
  const resp = await session.delete(`${GITEA_URL}/api/v1/teams/${team_id}/members/${username}`);
  return [resp.status, await resp.text()];
}

export async function add_team_repo(session: Session, team_id: number, course_name: string, repo_name: string): Promise<[number, string]> {
  const resp = await session.put(`${GITEA_URL}/api/v1/teams/${team_id}/repos/${course_name}/${repo_name}`);
  return [resp.status, await resp.text()];
}

// --- Repos ---

export async function create_blank_repo(session: Session, owner: string, repo_name: string): Promise<[string, unknown]> {
  const resp = await session.post(`${GITEA_URL}/api/v1/orgs/${owner}/repos`, {
    name: repo_name, private: true, auto_init: true,
  });
  if (resp.status === 201) return ['created', await resp.json()];
  return ['error', await resp.text()];
}

export async function ensure_repo(session: Session, template_owner: string, template_repo: string, new_owner: string, new_repo: string): Promise<[string, unknown]> {
  const resp = await session.get(`${GITEA_URL}/api/v1/repos/${new_owner}/${new_repo}`);
  if (resp.status === 200) {
    const repo = await resp.json();
    if (!repo.private) {
      const patch = await session.patch(`${GITEA_URL}/api/v1/repos/${new_owner}/${new_repo}`, { private: true });
      if (patch.status === 200) console.log(`  FIXED: Repo '${new_owner}/${new_repo}' set to private.`);
      else console.log(`  WARN: Could not set repo to private (HTTP ${patch.status})`);
    }
    return ['exists', repo];
  }
  const gen = await session.post(`${GITEA_URL}/api/v1/repos/${template_owner}/${template_repo}/generate`, {
    owner: new_owner, name: new_repo, private: true, git_content: true, topics: true, labels: true,
  });
  if (gen.status === 201) return ['created', await gen.json()];
  return ['error', await gen.text()];
}

export async function get_org_repos(session: Session, course_name: string): Promise<Record<string, unknown>[]> {
  const repos: Record<string, unknown>[] = [];
  let page = 1;
  while (true) {
    const resp = await session.get(`${GITEA_URL}/api/v1/orgs/${course_name}/repos`, { page, limit: 50 });
    if (resp.status !== 200) break;
    const data = await resp.json();
    if (!data.length) break;
    repos.push(...data);
    page++;
  }
  return repos;
}

// --- CSV file resolution ---

export function resolve_csv(course_name: string): string {
  return path.join(COURSES_DIR, `${course_name}.csv`);
}

export function parse_csv(file_path: string): CsvEntry[] {
  const content = fs.readFileSync(file_path, 'utf-8');
  const entries: CsvEntry[] = [];
  for (const line of content.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const parts = trimmed.split(/\s+/);
    entries.push([parts[0], parts[1], parts.length >= 4 ? parts[3] : null]);
  }
  return entries;
}

export function parse_csv_teams(file_path: string): [string | null, Map<string, string[]>] {
  const content = fs.readFileSync(file_path, 'utf-8');
  const course_names = new Set<string>();
  const teams = new Map<string, string[]>();

  for (const line of content.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const parts = trimmed.split(/\s+/);
    if (parts.length < 4) continue;
    course_names.add(parts[0]);
    if (!teams.has(parts[3])) teams.set(parts[3], []);
    teams.get(parts[3])!.push(parts[1]);
  }

  if (course_names.size > 1) {
    console.log(`Error: Multiple courses found in CSV: ${[...course_names].sort().join(', ')}`);
    console.log('  All rows must belong to the same course.');
    process.exit(1);
  }

  if (!course_names.size) return [null, teams];
  return [[...course_names][0], teams];
}

export function validate_single_course(entries: CsvEntry[]): string {
  const course_names = new Set(entries.map(([c]) => c));
  if (course_names.size > 1) {
    console.log(`Error: Multiple courses found in CSV: ${[...course_names].sort().join(', ')}`);
    console.log('  All rows must belong to the same course.');
    process.exit(1);
  }
  return [...course_names][0];
}
